src/main_functions.c

```diff
@@ -0,0 +1,606 @@
+			int insertcheck = 0;
+
+			// process of accepting replacment inputs
+			while (!userconsent) {
+				printf("\nEnter replacement: ");
+				insertcheck = scanf("%s", c);
+
+
+				while (insertcheck < 0) {
+					shell_error("\n\nPlease enter a valid input!\n");
+					shell_prompt();
+					insertcheck = scanf("%s\n", c);
+				}
+
+				printf("Are you sure you wish to replace \"%s\" with \"%s\"? [y, n] : ", misspelled[i], c);
+				userconsent = scanf("%s", sig);
+				if (sig[0] == 'y') {
+					userconsent = 1;
+				} else {
+					userconsent = 0;
+				}
+			}
+
+			char* newword = strdup(c);
+			printf("\nReplacing \"%s\" with \"%s\". \n", misspelled[i], newword);
+			correct_line(line_copy, misspelled[i], newword); //modifies line function
+
+         	printf("%s", line_copy);
+        	/* if the line is the last line being edited, the last character is
+        	 * EOF, not \n. */
+         	if (returnflag) printf("\n");
+         	printf("%s", underline_misspelled_sentence(misspelled, line_copy, underline));
+
+        } else if (isdigit(choice[0]) && (atoi(&choice[0]) <= max_no_suggestions)) { // choose suggestion
+        	// Replace misspelled word with chosen replacement
+        	int c = atoi(&choice[0]) - 1;
+        	printf("\nReplacing %s with %s.\n\n", misspelled[i], suggestions[c]);
+        	correct_line(line_copy, misspelled[i], suggestions[c]); //modifies line function
+
+        	/* if the line is the last line being edited, the last character is
+        	 * EOF, not \n. */
+         	printf("%s", line_copy);
+         	if (returnflag) printf("\n");
+         	printf("%s", underline_misspelled_sentence(misspelled, line_copy, underline));
+        }
+
+        i++;	// added loop changer
+    }
+
+	return line_copy;
+}
+
+
+char** interactive_mode(char* filename, dict_t* dict, int* quit)
+{
+	char** lines;
+	lines = parse_file(filename);
+
+	// step through phases
+	int i = 0;
+	int flag = 0;
+	int linenumber;
+
+	while (lines[i] != NULL) {
+		linenumber = i+1;
+		if (lines[i+1] == NULL) flag = 1;	// last line
+		lines[i] = edit_interactive(lines[i], dict, linenumber, flag);
+		i++;
+	}
+
+    *quit = 0;
+	return lines;
+}
+/* 
+	IV. Batch Mode
+ */
+// Adapted from sarika's edit_interactive
+// returns corrections to a list that contains misspelled words, or returns a corrected char** string
+// that can represent an array
+char* edit_batch(char* line, dict_t* dict, int verbosity, int lnum)
+{
+
+	char *line_copy = strdup(line);
+    int max_no_suggestions = 2; //should the user decide this?
+    int length = strlen(line);
+    char **misspelled; //generates an empty array where the misspelled words in a line will be stored
+
+    misspelled = calloc(length, sizeof(char*));
+    if (misspelled == NULL) {
+    	fprintf(stderr,"edit_interactive: calloc failed");
+    	exit(0);
+    }
+
+    char *underline = (char *)malloc(sizeof(char) * strlen(line + 1)); //generate an empty array where the underline will go
```

Comment:
Hey, this is still wrong. You're doing pointer arithmetic on line, which is bad since we'll get `strlen(line) - 1` instead. It should be `char *underline = (char *)malloc(sizeof(char) * (strlen(line) + 1))`. 

I'm not sure how these changes keep getting reverted but this is not the first one I've caught

Reply:
No, it's still there